### ISSUE TITLE
`@unocss/webpack` Can't compile styles in RSC(`React Server Components`) in Next.js 13

### ISSUE THREAD

**Issue by greenhat616** [unocss/unocss#3198]
### UnoCSS version

0.56.5

### Describe the bug
![image](https://github.com/unocss/unocss/assets/41122242/460b42ab-e0ec-40be-845a-1941bec0c4ef)

`@unocss/webpack` could not scan React Server Components and compile the related styles.

**Updated**: transformers **works** _as normal_, but the styles can't be compiled.

Related: <https://react.dev/blog/2022/03/29/react-v18#server-components-experimental>

## Configration

`uno.config.ts`
```ts
import {
  defineConfig,
  presetUno,
  presetIcons,
  presetAttributify,
} from 'unocss';
import presetRemToPx from '@unocss/preset-rem-to-px';
import transformerCompileClass from '@unocss/transformer-compile-class';
import transformerVariantGroup from '@unocss/transformer-variant-group';
import transformerDirectives from '@unocss/transformer-directives';
import transformerAttributifyJsx from '@unocss/transformer-attributify-jsx';

export default defineConfig({
  content: {
    pipeline: {
      // include: [/src\/.*\.(s?css|[jt]sx?)$/],
      include: [/\.([jt]sx|mdx?|html|s?css)($|\?)/],
      exclude: [],
    },
  },
  presets: [
    presetUno({
      dark: {
        dark: '.dark-mode',
        light: '.light-mode',
      },
    }),
    presetIcons(),
    presetAttributify(),
    presetRemToPx(),
  ],
  transformers: [
    transformerDirectives({
      enforce: 'pre',
    }),
    transformerCompileClass(),
    transformerVariantGroup(),
    transformerAttributifyJsx(),
  ],
});

```

`next.config.js`
```js
const UnoCSS = require('@unocss/webpack').default;

/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack(config) {
    // UnoCSS Support
    config.plugins.push(UnoCSS());
    return config;
  },
};

module.exports = nextConfig;
```

### Reproduction

See here: https://stackblitz.com/edit/stackblitz-starters-uirmkr?file=app%2Fpage.tsx

### System Info

Ubuntu 22.04 LTS
Chrome Latest
Next.js 13 with App Dirs and webpack bundled

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by antfu** [comment]
I personally have zero experience with React Server Component, and no really much interest in it. Open to contributions as long as the solution is straightforward or easy to maintain in the long term.

**Comment by henrikvilhelmberglund** [comment]
It mostly works with `@unocss/postcss` but `transformerAttributifyJsx()` breaks instead. Maybe a better option for now!

**Comment by greenhat616** [comment]
> It mostly works with `@unocss/postcss` but `transformerAttributifyJsx()` breaks instead. Maybe a better option for now!

No, PostCSS can't transform the code in files like tsx directly, to my understanding. So `complie classes transformer` and `transformerVariantGroup`(https://github.com/unocss/unocss/issues/3192) is not work yet.

**Comment by flyingalex** [comment]
I guess cause React Server Component already becomes to backend file, we need to use backend way
ref: https://unocss.dev/integrations/cli

```js
export default defineConfig({
  cli: {
    entry: {
      patterns: ['app/**/*.tsx'],
      outFile: 'app/uno-cli.css',
    }, // CliEntryItem | CliEntryItem[]
  },
  ....
}
```
then import ` 'app/uno-cli.css'` instead of `uno.css`, which seems work, but this way for me looks not very right

**Comment by quangkhaidam93** [comment]
> I guess cause React Server Component already becomes to backend file, we need to use backend way ref: https://unocss.dev/integrations/cli
> 
> ```js
> export default defineConfig({
>   cli: {
>     entry: {
>       patterns: ['app/**/*.tsx'],
>       outFile: 'app/uno-cli.css',
>     }, // CliEntryItem | CliEntryItem[]
>   },
>   ....
> }
> ```
> 
> then import ` 'app/uno-cli.css'` instead of `uno.css`, which seems work, but this way for me looks not very right

As fas as I investigate, `unocss/webpack` in NextJS ignores server components and does not create corresponding utility classes in final css in development mode. In production mode (npm run build), both NextJS and `uno.css/webpack` do not create corresponding utility classes with `transformerVariantGroup` (either server or client components)

**Comment by paulm17** [comment]
I don't seem to have these issues.  Both client and server components work and I get styles with the build.  Also the variant transformer works fine.

I do have an issue, that when I update the styles.  There is no refresh with the HMR.  I have to exit out of the process and restart for styles to take effect.

Can someone confirm they get the same for the styles?  

Thanks!

**Comment by paulm17** [comment]
I forked the webpack plugin specifically for nextjs: https://pastebin.com/raw/Pdf1GgqL

The key takeaway is that the webpack plugin runs in cycles where only certain files are available and also for the different server and client trees.  The second is a guess on my part (in relation to webpack).

Why am I saying this?  Prior to the fork, when running `use client`, the layout file has all the tokens needed for the styling.  Without `use client` there are no tokens present and therefore the time when the stylesheet needs to be overwritten only the preflight is available.

-----

How did I overcome this?  Simply put, I am collating any and all stying pertaining to the pages, saving to a file.  Then finding the layout.css file and overwriting it with the preflight and styles from the file.  I am using a file as an interlocutor as an ordinary variable becomes undefined in the process. 

This works on server and client pages, also includes the build process.

I've tested this only in the confines of the unocss repo.  I'm not making a PR due to not having the capacity to test against real-world scenarios.

Would appreciate if anyone wants to do the same and report back.  Thanks!
